### ISSUE TITLE
[UX] Password strength: provide suggestions to make it stronger

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5220]
This is a follow-up to #5095, where we've done a great job:

<img width=70% src="https://user-images.githubusercontent.com/2423362/133269849-abf7d18c-5341-4d88-bdee-87f9a9411bd6.png">

- the first bullet point does a good job, as it tells people what the exact expectation/requirement is 👍🏼
- the "long" requirement of the second bullet point is also straight-forward (user things "I need to type more"), but combining it with the word "sufficiently" makes it vague ("How many characters is long enough?") 👎🏼 
- the "complex" requirement in the second bullet point is also vague ("what makes a password complex enough ...what do I need to do?") 👎🏼 

I find some of the wording Drupal uses helpful:
<img width=60% src="https://user-images.githubusercontent.com/2423362/133270895-080a7fd2-0f47-451d-b903-e5bedb9ff603.png">

We could append something like this in the second bullet point:
> Try a mix of uppercase and lowercase letters, numbers, and punctuation.

**Comment by klonos** [comment]
> That's why the sentence "The password must be sufficiently long or complex" describes it well...

Fair enough= 👍🏼  ...let's give it some time, so that other can chime in. If no agreement or interest in this, then let's close it.

**Comment by herbdool** [comment]
I'm pretty sure we hashed out the wording in a previous issue. This was the compromise.

**Comment by stpaultim** [comment]
I'm fine with the current wording.

**Comment by klonos** [comment]
Thanks for taking the time to chime in everyone 👍🏼 ...closing then.

**Comment by stpaultim** [comment]
FYI - See this forum post. https://forum.backdropcms.org/forum/password-parameters

"Can anyone tell me what the actual minimum requirements are please? I can then change the string to better inform my users who are  non-native English speakers that will have a poor grasp of English. I envisage problems for my users who might well go elsewhere if they cannot register on my site if they cannot figure out the password criteria."

**Comment by robnacnamha** [comment]
It was I who posted on the forum to which @stpaultim refers and I would like to add here what functionality I would like to have:
Something like the password policy module in Drupal 7 that was configurable and the admin could choose the complexity etc that they wished, setting how many letters, cases, numbers and characters they wanted the user to include and the module helpfully told the user what was  still required if the criteria was not met.

I have looked at the passphrase policy module for backdrop which looks similar to the password policy but it does not appear to be configurable.

My previous job was managing a busy Drupal 7 site that I  had to introduce password policy and increase the security after  user's accounts were getting hacked due to using ridiculously insecure passwords. I then had a backlash and constant  grumbles from users that the security was frustratingly high and we lost subscribers and income because of it and put off  others from joining so I compromised and met them halfway with the  security criteria. So I  know from experience that I need to be able to set the level of security I want from my users.

Cheers
Rob

**Comment by indigoxela** [comment]
Hi @robnacnamha,

you can configure password strength handling in two ways:

1. By default the password strength is only a recommendation, so although the indicator's still showing red or yellow, the weak password is still accepted by Backdrop (same as in Drupal btw). You can switch that to "required" on "Account settings" (admin/config/people/settings, section "Password policy")
2. There's a hidden setting "user_password_strength_threshold" to loosen or further strengthen the password requirements in file system.core.json - default value 50. There's currently no user interface for that setting (yet)

**Comment by herbdool** [comment]
As per @indigoxela second bullet, it can be overridden in the settings.php file by adding something like:

```
$config['system.core']['user_password_strength_threshold'] = 80;
```

Looks like https://github.com/backdrop-contrib/passphrase_policy has no maintainer currently and that it was forked from the Drupal 7 version with the configuration stripped out, which is unfortunate. If someone wanted to become the maintainer they could also decide to put the configuration back in (copying from D7).

Anyway, *issue* is just about communicating the core password policy, and as discussed above, we are emphasizing *length* over specific mixes of symbols, letters and numbers. The indicator gives instant feedback so this is even more helpful than trying to nail down help text.

Another good reason to keep it vague is that trying to explain the actual calculation happening behind the scenes would probably confuse most people. For instance, we'd have to dive into concepts like "entropy" and "complexity". Even saying "The password must be sufficiently long or complex" is not exactly true. In fact, the length is part of the calculation to determine the complexity. So longer is always more complex.

**Comment by JugglingCoder** [comment]
Just my 2 cents.  I must agree with "robnacnamha" above.  I'd say most of my new users get extremely frustrated at what exactly they need to input as a password.  I want to keep my site(s) with the setting "Reject weak passwords" but the "everyday Joe" type person has no idea why the password they type in won't be accepted.  Most people just use letters (maybe with CAPS) and without specific instruction they keep trying and trying... can't get it to accept and say Adios.

A played around a lot with different types of passwords and from what I can tell (when using "Reject weak passwords"), this textual explanation is working for me:

At least 15 letters OR 11 or more characters that contains at least 1 "Special" Character (#,!,$,%, etc) and at least 1 Capitalized Character.  Optionally, add numbers too if you like.

This is probably something the developers don't want to add to core but it would be nice for site admins to be able to "alter" the explanation.  I see the text in "core/modules/user/user.password.inc > function _user_password_policy_help()" but of course I don't want to alter that because core files get over-written on updates but there is no hook_ either....

**Comment by indigoxela** [comment]
> it would be nice for site admins to be able to "alter" the explanation

That's an interesting suggestion. :thinking: Could we make it alterable? Or could we make this part of the config - and translatable?

**Comment by herbdool** [comment]
If someone wants to create a PR to add a hook we can consider it. Or a PR to add more suggestions of how to make a password stronger.

**Comment by avpaderno** [comment]
*The password must be sufficiently long or complex* could mean that *11111111111111111111111111111* is a good password.

I do not think there are criteria for passwords that would be good for all the sites, though. This seems a feature for a contributed module people could install if they need to put more restrictions on passwords.

**Comment by argiepiano** [comment]
>  most of my new users get extremely frustrated at what exactly they need to input as a password. 

I've seen this, too. I have gotten complaints about the lack of clarity in "The password must be sufficiently long or complex."

> This is probably something the developers don't want to add to core but it would be nice for site admins to be able to "alter" the explanation

Yes, it's simple to provide more information. Simply implement `hook_form_user_pass_reset_form_alter()` and do something like:

```php
  $form['pass']['#description'] = '<div class="password-strength-help-text fair">The password strength indicator must be <em class="placeholder">fair</em> or better. The following criteria must be met:</div><div class="item-list"><ul><li class="odd first">The password cannot be the same as the username or email address.</li><li class="even last">The password must be at least 11 characters long, and contain at least one uppercase letter, or one number, or one special character.</li></ul></div>';
```

**Comment by avpaderno** [comment]
I would like to see #4603 implemented.

**Comment by herbdool** [comment]
Right, someone can use a form alter hook. No PRs by anyone to make this easier yet.

> ...contain at least one uppercase letter, or one number, or one special character.

Someone could put that into an alter hook, though it isn't strictly true in regards to the current password strength algorithm.
